src/nnet3bin/nnet3-combine.cc

```diff
@@ -94,24 +152,45 @@ int main(int argc, char *argv[]) {
       KALDI_ASSERT(!egs.empty());
     }
 
+    // first evaluates the objective using the last model.
+    int32 best_num_to_combine = 1;
+    double best_objf = ComputeObjf(batchnorm_test_mode, dropout_test_mode,
+        egs, moving_average_nnet, &prob_computer);
+    KALDI_LOG << "objective function using the last model is " << best_objf;
 
     int32 num_nnets = po.NumArgs() - 2;
-    if (num_nnets > 1 || !combine_config.enforce_sum_to_one) {
-      NnetCombiner combiner(combine_config, num_nnets, egs, nnet);
-
+    // then each time before we re-evaluate the objective function, we will add
+    // num_to_add models to the moving average.
+    int32 num_to_add = (num_nnets + max_objective_evaluations - 1) /
+                       max_objective_evaluations;
+    if (num_nnets > 1) {
       for (int32 n = 1; n < num_nnets; n++) {
         ReadKaldiObject(po.GetArg(1 + n), &nnet);
-        combiner.AcceptNnet(nnet);
+        // updates the moving average
+        UpdateNnetMovingAverage(n + 1, nnet, &moving_average_nnet);
+        // evaluates the objective everytime after adding num_to_add model or
+        // all the models to the moving average.
+        if ((n - 1) % num_to_add == num_to_add - 1 || n == num_nnets - 1) {
+          double objf = ComputeObjf(batchnorm_test_mode, dropout_test_mode,
+              egs, moving_average_nnet, &prob_computer);
+          KALDI_LOG << "Combining last " << n + 1
+                    << " models, objective function is " << objf;
+          if (objf > best_objf) {
+            best_objf = objf;
+            best_nnet = moving_average_nnet;
+            best_num_to_combine = n + 1;
+          }
+        }
       }
-      combiner.Combine();
+      KALDI_LOG << "Using the model averaged over last " << best_num_to_combine
+                << " models, objective function is " << best_objf;
 
 #if HAVE_CUDA==1
       CuDevice::Instantiate().PrintProfile();
 #endif
-      nnet = combiner.GetNnet();
       if (HasBatchnorm(nnet))
-        RecomputeStats(egs, &nnet);
-      WriteKaldiObject(nnet, nnet_wxfilename, binary_write);
+        RecomputeStats(egs, &best_nnet);
+      WriteKaldiObject(best_nnet, nnet_wxfilename, binary_write);
     } else {
       KALDI_LOG << "Copying the single input model directly to the output, "
```

Comment:
I believe this branch is now longer necessary, you can remove the if-statement and treat it like the other case, which will simplify the parsing the diagnostics.  (probably the branch was added to prevent some kind of crash).